src/listpack.c

```diff
@@ -647,8 +673,100 @@ static inline unsigned char *lpGetWithSize(unsigned char *p, int64_t *count, uns
     }
 }
 
+/* Return the listpack element pointed by 'p'.
+ *
+ * The function has the same behaviour as lpGetWithSize when 'entry_size' is NULL,
+ * but avoids a lot of unecesarry branching performance penalties.
+ * */
+static inline unsigned char *lpGetWithBuf(unsigned char *p, int64_t *count, unsigned char *intbuf) {
+    int64_t val;
+    uint64_t uval, negstart, negmax;
+    const unsigned char encoding = p[0];
+
+    assert(p); /* assertion for valgrind (avoid NPD) */
+    /* string encoding */
+    if (LP_ENCODING_IS_6BIT_STR(encoding)) {
+        *count = LP_ENCODING_6BIT_STR_LEN(p);
+        return p+1;
+    }
+    if (LP_ENCODING_IS_12BIT_STR(encoding)) {
+        *count = LP_ENCODING_12BIT_STR_LEN(p);
+        return p+2;
+    }
+    if (LP_ENCODING_IS_32BIT_STR(encoding)) {
+        *count = LP_ENCODING_32BIT_STR_LEN(p);
+        return p+5;
+    }
```

Comment:
does there advance string `if` cause regression for small int?
in the old code it wouldn't go into these two if blocks in these cases.

Reply:
@sundb it does not seem like it. I've validated with a test specifically focusing on int enconding (across all possibilities).

```

redis-cli LPUSH list:100:ints 0 7 14 21 28 35 42 49 56 63 70 77 84 91 98 105 112 -4096 -3585 -3074 -2563 -2052 -1541 -1030 -519 -8 503 1014 1525 2036 2547 3058 3569 4080 -32768 -28673 -24578 -20483 -16388 -12293 -8198 -4103 -8 4087 8182 12277 16372 20467 24562 28657 32752 -8388608 -7340033 -6291458 -5242883 -4194308 -3145733 -2097158 -1048583 -8 1048567 2097142 3145717 4194292 5242867 6291442 7340017 8388592 -2147483648 -1861152495 -1574821342 -1288490189 -1002159036 -715827883 -429496730 -143165577 143165576 429496729 715827882 1002159035 1288490188 1574821341 1861152494 2147483647 -9223372036854775808 -7993589098607472367 -6763806160360168926 -5534023222112865485 -4304240283865562044 -3074457345618258603 -1844674407370955162 -614891469123651721 614891469123651720 1844674407370955161 3074457345618258602 4304240283865562043 5534023222112865484 6763806160360168925 7993589098607472366 9223372036854775807
(integer) 100
```

and then benchmarked 

unstable (06b144aa097e968f579048800cd0e5c9336ea1c9)

```
root@hpe10:~/redis# taskset -c 2-5 memtier_benchmark -S /tmp/1.socket --pipeline 10 --command="LRANGE list:100:ints 0 -1"  --hide-histogram --test-time 60
Writing results to stdout
[RUN #1] Preparing benchmark client...
[RUN #1] Launching threads now...
[RUN #1 100%,  60 secs]  0 threads:     7388830 ops,  122924 (avg:  123121) ops/sec, 172.56MB/sec (avg: 172.84MB/sec), 16.10 (avg: 16.12) msec latency

4         Threads
50        Connections per thread
60        Seconds


ALL STATS
==================================================================================================
Type         Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
--------------------------------------------------------------------------------------------------
Lranges    123121.80        16.11794        16.19100        19.71100        22.91100    176987.58 
Totals     123121.80        16.11794        16.19100        19.71100        22.91100    353975.16 
```

this PR (d6799c6435d85c12e81e970a486689b7557a7e9d)

```
root@hpe10:~/redis# taskset -c 2-5 memtier_benchmark -S /tmp/1.socket --pipeline 10 --command="LRANGE list:100:ints 0 -1"  --hide-histogram --test-time 60
Writing results to stdout
[RUN #1] Preparing benchmark client...
[RUN #1] Launching threads now...
[RUN #1 100%,  60 secs]  0 threads:     8890670 ops,  147770 (avg:  148148) ops/sec, 207.44MB/sec (avg: 207.97MB/sec), 13.39 (avg: 13.37) msec latency

4         Threads
50        Connections per thread
60        Seconds


ALL STATS
==================================================================================================
Type         Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec 
--------------------------------------------------------------------------------------------------
Lranges    145719.97        13.37421        13.37500        18.17500        23.16700    209472.46 
Totals     145719.97        13.37421        13.37500        18.17500        23.16700    418944.91 
```

Nonetheless, let me keep this discussion open for me to add 2 tests to CI:
- the one above with 100 int elements, listpack encoded
- another one just for LP_ENCODING_IS_7BIT_UINT, which should show us the worst case scenario

Reply:
@sundb I've added 2 new benchmarks, one focusing on the "worst-case" change and we still see 11.7% and 14.3% improvements on the ops/sec:

Test Name | Metric JSON Path | Baseline unstable branch (06b144aa097e968f579048800cd0e5c9336ea1c9) | This PR (d6799c6435d85c12e81e970a486689b7557a7e9d) | % change
-- | -- | -- | -- | --
memtier_benchmark-1key-list-100-elements-int-7bit-uint-lrange-all-elements-pipeline-10 | ALL STATS.Totals."Ops/sec" | 71304.1 | 81531.75 | 14.3%
memtier_benchmark-1key-list-100-elements-int-7bit-uint-lrange-all-elements-pipeline-10 | ALL STATS.Totals."Percentile Latencies"."p50.00" | 2.183 | 1.919 | 12.1%
memtier_benchmark-1key-list-100-elements-int-7bit-uint-lrange-all-elements-pipeline-10 | ALL STATS.Totals."Percentile Latencies"."p99.00" | 4.311 | 3.783 | 12.2%
memtier_benchmark-1key-list-100-elements-int-lrange-all-elements-pipeline-10 | ALL STATS.Totals."Ops/sec" | 62213.7 | 69474.09 | 11.7%
memtier_benchmark-1key-list-100-elements-int-lrange-all-elements-pipeline-10 | ALL STATS.Totals."Percentile Latencies"."p50.00" | 2.487 | 2.247 | 9.7%
memtier_benchmark-1key-list-100-elements-int-lrange-all-elements-pipeline-10 | ALL STATS.Totals."Percentile Latencies"."p99.00" | 4.943 | 4.447 | 10.0%

I'll trigger a full benchmark run and these 2 benchmarks will enter the "official" tables.